drivers/video/video_esp32_dvp.c

```diff
@@ -0,0 +1,463 @@
+		LOG_ERR("No video buffer available. Enque some buffers first.");
+		return -EAGAIN;
+	}
+
+	ret = dma_reload(cfg->dma_dev, cfg->rx_dma_channel, 0, (uint32_t)data->active_vbuf->buffer,
+			 data->active_vbuf->bytesused);
+	if (ret) {
+		LOG_ERR("Unable to reload DMA (%d)", ret);
+		return ret;
+	}
+
+	ret = dma_start(cfg->dma_dev, cfg->rx_dma_channel);
+	if (ret) {
+		LOG_ERR("Unable to start DMA (%d)", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int video_esp32_stream_start(const struct device *dev)
+{
+	const struct video_esp32_config *cfg = dev->config;
+	struct video_esp32_data *data = dev->data;
+	struct dma_status dma_status = {0};
+	struct dma_config dma_cfg = {0};
+	struct dma_block_config *dma_block_iter = data->dma_blocks;
+	uint32_t buffer_size = 0;
+	unsigned int key = irq_lock();
+	int error = 0;
+
+	LOG_DBG("Start streaming");
+
+	error = dma_get_status(cfg->dma_dev, cfg->rx_dma_channel, &dma_status);
+
+	if (error) {
+		LOG_ERR("Unable to get Rx status (%d)", error);
+		goto unlock;
+	}
+
+	if (dma_status.busy) {
+		LOG_ERR("Rx DMA Channel %d is busy", cfg->rx_dma_channel);
+		error = -EBUSY;
+		goto unlock;
+	}
+
+	data->active_vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+	if (!data->active_vbuf) {
+		LOG_ERR("No enqueued video buffers available.");
+		error = -EAGAIN;
+		goto unlock;
+	}
+
+	buffer_size = data->active_vbuf->bytesused;
+	memset(data->dma_blocks, 0, sizeof(data->dma_blocks));
+	for (int i = 0; i < CONFIG_DMA_ESP32_MAX_DESCRIPTOR_NUM; ++i) {
+		dma_block_iter->dest_address =
+			(uint32_t)data->active_vbuf->buffer + i * VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+		if (buffer_size < VIDEO_ESP32_DMA_BUFFER_MAX_SIZE) {
+			dma_block_iter->block_size = buffer_size;
+			dma_block_iter->next_block = NULL;
+			dma_cfg.block_count = i + 1;
+			break;
+		}
+		dma_block_iter->block_size = VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+		dma_block_iter->next_block = dma_block_iter + 1;
+		dma_block_iter++;
+		buffer_size -= VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+	}
+
+	if (dma_block_iter->next_block) {
+		LOG_ERR("Not enough descriptors available. Increase DMA_ESP32_DESCRIPTOR_NUM");
+		return -ENOBUFS;
+	}
+
+	dma_cfg.channel_direction = PERIPHERAL_TO_MEMORY;
+	dma_cfg.dma_callback = video_esp32_dma_rx_done;
+	dma_cfg.user_data = data;
+	dma_cfg.dma_slot = SOC_GDMA_TRIG_PERIPH_CAM0;
+	dma_cfg.complete_callback_en = 1;
+	dma_cfg.head_block = &data->dma_blocks[0];
+
+	error = dma_config(cfg->dma_dev, cfg->rx_dma_channel, &dma_cfg);
+	if (error) {
+		LOG_ERR("Unable to configure DMA (%d)", error);
+		goto unlock;
+	}
+
+	LCD_CAM.cam_ctrl1.cam_reset = 1;
+	LCD_CAM.cam_ctrl1.cam_reset = 0;
+	LCD_CAM.cam_ctrl1.cam_afifo_reset = 1;
+	LCD_CAM.cam_ctrl1.cam_afifo_reset = 0;
+
+	error = dma_start(cfg->dma_dev, cfg->rx_dma_channel);
+	if (error) {
+		LOG_ERR("Unable to start DMA (%d)", error);
+		goto unlock;
+	}
+
+	LCD_CAM.cam_ctrl.cam_update = 1;
+	LCD_CAM.cam_ctrl.cam_update = 0;
+	LCD_CAM.cam_ctrl1.cam_start = 1;
+
+	data->is_streaming = true;
```

Comment:
Question, is it a must to place a giant lock inside of this function?

Reply:
probably not, since the interrupts are inactive when you start the stream any how...
I'll remove it and add a check to make sure we are not already streaming on stream_start().